### PR TITLE
hw: m25p80: use upstream suggestions -  move write_enable to decode_n…

### PR DESCRIPTION
**Pull request by chen-iris**
…ew_cmd

Match suggestions given by https://lore.kernel.org/qemu-devel/20220609192201.GC10629@fralle-msi/
Major changes: 
- Moved the write_enable `wp_level` check to `decode_new_cmd`  based on suggestions from the upstream
- Added subsections

Signed-off-by: Iris Chen <irischenlj@fb.com>

### CODE REVIEW

**Review by peterdelevoryas — APPROVED** [review]
I think this is right, and it's pretty simple actually, nice job. It's gonna be a nightmare when I have to rebase all these changes and merge with the upstream and remember what's what.

File: hw/block/m25p80.c [hunk 1]
```diff
@@ -1670,6 +1671,43 @@ static const VMStateDescription vmstate_m25p80_aai_enable = {
...
+
+    return s->status_register_write_disabled;
+}
+
+static const VMStateDescription
+             vmstate_m25p80_status_register_write_disabled = {
+    .name = "m25p80/status_register_write_disabled",
+    .version_id = 1,
+    .minimum_version_id = 1,
+    .needed = vmstate_m25p80_status_register_write_disabled_needed,
```
Comment (peterdelevoryas) [inline hw/block/m25p80.c]: I don't think we should include a .needed function, we should just leave it false.

File: hw/block/m25p80.c [hunk 2]
```diff
@@ -1670,6 +1671,43 @@ static const VMStateDescription vmstate_m25p80_aai_enable = {
...
+{
+    Flash *s = (Flash *)opaque;
+
+    return s->status_register_write_disabled;
+}
+
+static const VMStateDescription
+             vmstate_m25p80_status_register_write_disabled = {
+    .name = "m25p80/status_register_write_disabled",
+    .version_id = 1,
```
Comment (peterdelevoryas) [inline hw/block/m25p80.c]: I also think maybe the version_id should be 2, but not totally sure.

File: hw/block/m25p80.c [hunk 3]
```diff
@@ -1670,6 +1671,43 @@ static const VMStateDescription vmstate_m25p80_aai_enable = {
...
+static bool vmstate_m25p80_status_register_write_disabled_needed(void *opaque)
+{
+    Flash *s = (Flash *)opaque;
+
+    return s->status_register_write_disabled;
+}
+
+static const VMStateDescription
+             vmstate_m25p80_status_register_write_disabled = {
+    .name = "m25p80/status_register_write_disabled",
```
Comment (peterdelevoryas) [inline hw/block/m25p80.c]: We should put both vmstate attributes in a single subsection, instead of having two separate subsections.

**Review by peterdelevoryas — CHANGES_REQUESTED** [review]

**Comment by chen-iris** [comment]
Updated based on all suggestions: 
- Combine into one VMState: let me know if you have feedback with regards to the naming

**Pull request merged**